### PR TITLE
fix Document.objects.filter(*).explain(), plus a few fixes for flakey tests

### PR DESCRIPTION
**Pull request by alsuren**
If you want me to split this up into separate PRs then I can. Each of the commits should probably be reviewed independently though.

### CODE REVIEW

File: sample_test_settings.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
     'db_name': 'test_sharding'
 }
 CONN2 = {
-    'uri': 'mongodb://replica:27018/?replicaSet=cluster-2',
+    'uri': 'mongodb://replica:27017/?replicaSet=cluster-2',
```
Comment (iainb) [inline sample_test_settings.py]: iirc there are supposed to be two separate mongos to run the tests.

Comment (alsuren) [inline sample_test_settings.py]: Cookie changed it to `controller:27017` and `replica:27018` when he wrote the docker-compose file, but `replica` is actually running on the default port (27017) too.

On that note: does anyone know how to upload the new version of shardmonster/docs/source/developing.rst to http://shardmonster.readthedocs.org/en/latest/ ?

Comment (iainb) [inline sample_test_settings.py]: Oh I know why I'm getting confused. Colin originally ran the tests without docker/docker compose, there were several Mongo instances running on different ports within the VM, where `controller` / `replica` all resolved to localhost.

Whereas Cookie's tests runs the whole thing using docker compose, so the `replica` and `controller` aren't actually bound to the same port. 

Carry on! :+1:

You'd have to talk to Colin about updating readthedocs, there's also the package on pypi.

File: shardmonster/tests/test_operations.py [hunk 2]
```diff
@@ -483,7 +483,8 @@ def test_skip_slice(self):
 
         c = operations.multishard_find('dummy', {'y': 1})[1:]
         results = sorted(list(c), key=lambda d: d['x'])
-        self.assertEquals([doc2], results)
+        # we're not specifying a sort, so the order is undefined.
+        self.assertIn(results, [[doc1], [doc2]])
```
Comment (iainb) [inline shardmonster/tests/test_operations.py]: You can use `assertItemsEqual` to compare two lists where the ordering is unimportant.

Though upon further reading of the test that is not what you're doing anyway. Does this test actually test anything useful now? The other changes in this commit also seem to suffer from the same problem.

Comment (alsuren) [inline shardmonster/tests/test_operations.py]: they basically test that the length of the returned list is 1. I did a git blame before I changed the tests, and it looks like they're all catching problems where the returned list ends up being empty. I can go back and revert the bits of operations.py where the tests were introduced, and make double sure they are still useful tests.

**Comment by iainb** [comment]
re: 46553079ef6ee4047d542c34d2ba0238098d1b67 (honour skip) is this change the cause of the test flakes in later commits? 

I think some additional tests to confirm this behaviour works would be a sensible addition.

**Comment by iainb** [comment]
Last time I changed shard monster, to gain confidence I upgraded shard monster in the monolith and ran the test suite. I think you should do the same, it may shake out any bugs/changes that the shard monster tests haven't picked up.

**Comment by alsuren** [comment]
> re: 4655307 (honour skip) is this change the cause of the test flakes in later commits?

No. They were flakey before my change. I think that the order has always depended on the python dictionary iteration order, which tox randomises at startup. I suspect that the original author may have been running `setup.py nosetests` directly? An alternative solution would be to clamp PYTHONHASHSEED at a fixed value, or force `_create_collection_iterator()` to iterate over locations alphabetically. https://tox.readthedocs.io/en/latest/example/basic.html#special-handling-of-pythonhashseed

> I think some additional tests to confirm this behaviour works would be a sensible addition.

I suspect that alphabetical iteration over shard locations will make it easier to write test-cases that can be reasoned about. Does that sound like a sensible idea?

**Pull request closed without merging**